### PR TITLE
New IAS authentication

### PR DESCRIPTION
**Pull request by g2flyer**
- convert to new IAS authentication from specification release 5.0
- build and typo fix for selection of linkability patch

Solves issue #47

### CODE REVIEW

File: ercc/Makefile [hunk 1]
```diff
@@ -14,6 +14,33 @@ decorator-plugin:
...
+DOCKER_CONTAINER=$(PEER_NAME)-$(ERCC_NAME)-0
+DOCKER_IMAGE?=$$(docker images | grep -- -$(ERCC_NAME)- | awk '{print $$1;}')
+
+docker:
+	docker build -t $(DOCKER_IMAGE) .
+
+docker-run:
+	docker run \
+                -it \
+                --device /dev/isgx \
```
Comment (mbrandenburger) [inline ercc/Makefile]: why do we need access to sgx in ercc?

File: ercc/Makefile [hunk 2]
```diff
@@ -14,6 +14,33 @@ decorator-plugin:
...
+                -v /var/run/aesmd:/var/run/aesmd \
+                --name $(DOCKER_CONTAINER) \
+                -e "CORE_CHAINCODE_LOGGING_LEVEL=DEBUG" \
+                -e "CORE_CHAINCODE_LOGGING_SHIM=INFO" \
+                -e "CORE_PEER_ADDRESS=172.17.0.1:7051" \
+                -e "CORE_CHAINCODE_ID_NAME=ecc:0" \
```
Comment (mbrandenburger) [inline ercc/Makefile]: this should be ercc

File: ercc/Makefile (hunk 1 above)
Comment (mbrandenburger) [reply ercc/Makefile]: mhh ... just wondering why this docker build for ercc at all?

File: fabric/README.md [hunk 3]
```diff
@@ -47,26 +47,21 @@ The requirements are:
...
-the constant `SGX_UNLINKABLE_SIGNATURE` to `SGX_LINKABLE_SIGNATURE`,
-re-compile and re-deplay [ecc_enclave](../ecc_enclave#build) and [ecc](../ecc#getting-started),
-configure your IAS settings as above with your linkable credentials and run the auction example as follows.
+with Intel. [Here](https://software.intel.com/en-us/form/sgx-onboarding)
+you can find more details on how to register and optain your SPID plus corresponding api-key.
+
+We currently support both `linkable' and 'unlinkable' signatures for the attestation.
+The type of attestation used is selected based on your 'ECC_ATTESTATION_TYPE' environment variable:
+'epid_unlinkable' for unlinkable or 'epid_linkable' for linkable signatures. The default in case you 
+do not define that environment variable is 'epid_unlinkable'.
```
Comment (mbrandenburger) [inline fabric/README.md]: Maybe rewording? By default `ECC_ATTESTATION_TYPE` is set to `epid_unlinkable`

File: ercc/attestation/ias_test.go [hunk 4]
```diff
@@ -25,6 +25,9 @@ const enclavePK = `MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEE9lPD9QkW9oxWlFvwABrmseYA
...
 
+
+// TODO: below should be fixed but i have no idea where NewIASCredentialProviderFromConfig comes from and
```
Comment (mbrandenburger) [inline ercc/attestation/ias_test.go]: let me check

**Comment by bvavala** [comment]
> Nice PR! Finally we get rid of that client cert. I like it.
> 
> I can see some minors here but still want to run it. To get my API key I just need to re-register at IAS?

Yes, through the portal -- https://api.portal.trustedservices.intel.com/ , super-easy

File: ercc/Makefile (hunk 1 above)
Comment (g2flyer) [reply ercc/Makefile]: was copy/paste, so definitely no need for sgx.  Definitely wanted the docker-clean aspect of it (which i've also added to ecc) but thought for dev mode a docker run could be useful. But thinking of it it actually not really  useful. will remove

File: ercc/Makefile (hunk 2 above)
Comment (g2flyer) [reply ercc/Makefile]: is removed now (see above)

File: fabric/README.md (hunk 3 above)
Comment (g2flyer) [reply fabric/README.md]: we do not set the environment variable, just set a default build variable if env variable is absent. did rephrase though a bit

File: ercc/attestation/ias_test.go (hunk 4 above)
Comment (g2flyer) [reply ercc/attestation/ias_test.go]: i suspect it's dead code, at least everything seems to work without fixing this. I've rephrased the comment to specifically refer to this issue/pr, so there is no urge to fix this as with attestation refactoring this might become double obsolete?

File: ecc_enclave/sgxcclib/sgxcclib.h [hunk 5]
```diff
@@ -18,7 +18,6 @@
 #define _SGXCCLIB_H_
 
 #include "types.h"
-#include "sgx_attestation_type.h"
```
Comment (mbrandenburger) [inline ecc_enclave/sgxcclib/sgxcclib.h]: this is a good catch!

File: ercc/attestation/ias.go [hunk 6]
```diff
@@ -109,6 +111,8 @@ func (ias *intelAttestationServiceImpl) RequestAttestationReport(cert tls.Certif
 		return IASAttestationReport{}, fmt.Errorf("IAS connection error: %s", err)
 	}
 	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Ocp-Apim-Subscription-Key", apiKey)
+	logger.Warningf("Sending IAS request %s", req)
```
Comment (mbrandenburger) [inline ercc/attestation/ias.go]: mh ... I have the feeling this should not be a warning. Moreover, this shim logger is kind of broken. Maybe we can omit this logging statement at all. In particular, this has the benefit that we do not add a dependency of the shim.

WDYT?

File: fabric/README.md [hunk 7]
```diff
@@ -47,26 +47,21 @@ The requirements are:
...
-registering it and get a SPID.
-
-We currently make use of `unlinkable signatures` for the attestation, thus, when registering with the IAS please choose
-unlinkable signatures. In the case you prefer linkable attestation,
-e.g., because you already have linkable IAS EPID credentials, change
-in line 217 of [../ecc_enclave/sgxcclib/sgxcclib.c](../ecc_enclave/sgxcclib/sgxcclib.c)
-the constant `SGX_UNLINKABLE_SIGNATURE` to `SGX_LINKABLE_SIGNATURE`,
-re-compile and re-deplay [ecc_enclave](../ecc_enclave#build) and [ecc](../ecc#getting-started),
-configure your IAS settings as above with your linkable credentials and run the auction example as follows.
+with Intel. [Here](https://software.intel.com/en-us/form/sgx-onboarding)
```
Comment (mbrandenburger) [inline fabric/README.md]: Would it make sense also to link to this here  https://api.portal.trustedservices.intel.com/

File: ecc_enclave/sgxcclib/sgxcclib.h (hunk 5 above)
Comment (g2flyer) [reply ecc_enclave/sgxcclib/sgxcclib.h]: Yep, clearly neither you nor Bruno did a clean build of the other PR  ... ;-)

File: ercc/attestation/ias.go (hunk 6 above)
Comment (g2flyer) [reply ercc/attestation/ias.go]: oops, this is a relict of debugging logging. I've actually fixed it but overlooked that file to add for an amend.  Should be fixed now as Debugf

File: fabric/README.md (hunk 7 above)
Comment (g2flyer) [reply fabric/README.md]: i thought i fixed that but apparently didn't. anyway, did now (but linked directly to the epid one https://api.portal.trustedservices.intel.com/EPID-attestation as right now we are not yet supporting ecdsa/dcap)

File: ecc_enclave/sgxcclib/sgxcclib.h (hunk 5 above)
Comment (mbrandenburger) [reply ecc_enclave/sgxcclib/sgxcclib.h]: yeah! +1 for CI build

File: fabric/README.md (hunk 7 above)
Comment (mbrandenburger) [reply fabric/README.md]: mh ... I have registered for a linkable epid. Got a SPID and a primary and a secondary key? which one we I use here?

Comment (g2flyer) [reply fabric/README.md]: i always use primary, don't know whether there is any difference in secondary, never tried with it.  In any case, updated now the docu to mention to use the primary one (and removed some other now stale info on certs/private-keys which i forgot to fix before ...)

File: fabric/README.md [hunk 8]
```diff
@@ -42,31 +42,25 @@ for a simple consortium and a bunch of scripts to run the auction demo.
...
 code tries to verify the attestation. (Another cause for such error '400'
-could a mismatch between provided SPID and client key as specified below).
+could a mismatch between provided SPID and api key as specified below).
 
-Place your client certificate and your SPID in the ``ias`` folder.
+Place your ias api key and your SPID in the ``ias`` folder as follows:
 
-    cp client.crt ${GOPATH}/src/github.com/hyperledger-labs/fabric-secure-chaincode/fabric/sgxconfig/ias/client.crt
-    cp client.key ${GOPATH}/src/github.com/hyperledger-labs/fabric-secure-chaincode/fabric/sgxconfig/ias/client.key
+    echo 'YOUR_API_KEY' > ${GOPATH}/src/github.com/hyperledger-labs/fabric-secure-chaincode/fabric/sgxconfig/ias/client.key
```
Comment (mbrandenburger) [inline fabric/README.md]: this key should go into `ias/api_key.txt`

Comment (g2flyer) [reply fabric/README.md]: aargh, i'm sure i fixed that at some point in time, i fear i might have overwritten at some point one of the git commit amends via a force push :-(   But should be fixed now as well ...

File: ercc/ercc.go [hunk 9]
```diff
@@ -17,8 +17,9 @@
 package main
 
 import (
+	"strings"
+
```
Comment (mbrandenburger) [inline ercc/ercc.go]: did you run gofmt?

File: fabric/README.md [hunk 10]
```diff
@@ -42,31 +42,25 @@ for a simple consortium and a bunch of scripts to run the auction demo.
...
-in line 217 of [../ecc_enclave/sgxcclib/sgxcclib.c](../ecc_enclave/sgxcclib/sgxcclib.c)
-the constant `SGX_UNLINKABLE_SIGNATURE` to `SGX_LINKABLE_SIGNATURE`,
-re-compile and re-deplay [ecc_enclave](../ecc_enclave#build) and [ecc](../ecc#getting-started),
-configure your IAS settings as above with your linkable credentials and run the auction example as follows.
+* a Service Provider ID (SPID)
+* the (primary) api-key associated with your SPID
+
+In order to use Intel's Attestation Service (IAS), you need to register
+with Intel. On the [IAS EPID registration page](https://api.portal.trustedservices.intel.com/EPID-attestation)
+you can find more details on how to register and optain your SPID plus corresponding api-key.
```
Comment (mbrandenburger) [inline fabric/README.md]: obtain

File: ercc/ercc.go (hunk 9 above)
Comment (g2flyer) [reply ercc/ercc.go]: i thought make would do that for me ;-)
more seriously, running now gofmt -w ercc/ercc..go only changes white spaces in front of a comment, nothing related to import. anyway, run no gofmt on {ercc,ecc}/..../*.go which also changed some whitespaces at a few other places and in one case re-ordered import  ...

**Comment by mbrandenburger** [comment]
@bvavala WDYT? Approved?

**Comment by bvavala** [comment]
Let me test it first. And the purpose is not really "testing" but rather updating my environment to make it work.

A good point for @g2flyer and @mbrandenburger is to make sure we get everybody's approval for PR that require modification to the dev environmnet.

File: ercc/Makefile [hunk 11]
```diff
@@ -14,6 +14,11 @@ decorator-plugin:
...
 	go test -v
 
-clean:
+clean: docker-clean
 	go clean
 
+ERCC_NAME?=ercc
+DOCKER_IMAGE?=$$(docker images | grep -- -$(ERCC_NAME)- | awk '{print $$1;}')
+
+docker-clean: docker-stop
```
Comment (bvavala) [inline ercc/Makefile]: forgot the docker-stop target?

Comment (g2flyer) [reply ercc/Makefile]: yes and no. i did have that target but then marcus convinced me that for ercc it doesn't make sense to have docker-start and docker-stop and so removed it, but forgot to remove also this dependency ...

Comment (mbrandenburger) [reply ercc/Makefile]: ok; will take care in #52
